### ISSUE TITLE
`jnp.floor_divide()` returns incorrect results when used with `scan`

### ISSUE THREAD

**Issue by jakevdp** [google/jax#9604]
```python
import jax
from jax import lax
import jax.numpy as jnp

x = jnp.array([2, 2, -3])

def reduce_for_loop(func, arr):
  val = arr[0]
  for i in range(1, len(arr)):
    val = func(val, arr[i])
  return val

def reduce_scan(func, arr):
  val = arr[0]
  body_func = lambda i, val: func(val, arr[i])
  return lax.fori_loop(1, len(arr), body_func, val)

print(reduce_for_loop(jnp.floor_divide, x)) # prints -1
print(reduce_scan(jnp.floor_divide, x))     # prints 0
with jax.disable_jit():
  print(reduce_scan(jnp.floor_divide, x))   # prints -1
```
Came up when working on #9529

**Comment by mattjj** [comment]
@jakevdp should we assign this, or label it with a priority that indicates we're not going to assign it right away? (Assigning to you for now just so you can make that decision!)

**Comment by jakevdp** [comment]
This seems like it should be high priority (silent production of incorrect results is bad!) but I've spent a while investigating and I'm no closer to understanding the root of the issue. Any thoughts?

**Comment by jakevdp** [comment]
Marking as P1 because this is urgent, but I am not actively working on this because I have no idea how to approach it.

**Comment by shoyer** [comment]
I get the correct behavior when I run this on a Colab GPU, so this seems to be an XLA CPU bug (maybe also a TPU bug?)

**Comment by soraros** [comment]
Definitely a XLA CPU bug.
```python
def reduce_scan_lax(a):
  def f(x1, x2):
    q = (lax.div(x1, x2) / 1).astype(int)
    r = (lax.rem(x1, x2) == 0)
    ret = r - q + q
    return ret, ret
  return lax.scan(f, 2, a)

def reduce_scan_lax_bad(a):
  def f(x1, x2):
    q = lax.div(x1, x2)
    r = lax.rem(x1, x2) == 0
    ret =  r - q + q
    return ret, ret
  return lax.scan(f, 2, a)

x = jnp.array([2, -3])
```
`reduce_scan_lax` and `reduce_scan_lax_bad` generate almost the same compiled `HloModule` on CPU, but only `reduce_scan_lax_bad` is, well, bad. I think they only differ by a type conversion if you inline some fused computation.
And if something other than `ret, ret` is returned, the problem also goes away.

**Comment by mattjj** [comment]
This still repros! I used @soraros 's code, with these imports and print statements:

```python
import jax
from jax import lax
import jax.numpy as jnp

def reduce_scan_lax(a):
  def f(x1, x2):
    q = (lax.div(x1, x2) / 1).astype(int)
    r = (lax.rem(x1, x2) == 0)
    ret = r - q + q
    return ret, ret
  return lax.scan(f, 2, a)

def reduce_scan_lax_bad(a):
  def f(x1, x2):
    q = lax.div(x1, x2)
    r = lax.rem(x1, x2) == 0
    ret =  r - q + q
    return ret, ret
  return lax.scan(f, 2, a)

x = jnp.array([2, -3])

print(reduce_scan_lax(x))
print(reduce_scan_lax_bad(x))
```

**Comment by jakevdp** [comment]
Internal tracking bug: b/130349807

**Comment by jakevdp** [comment]
Internal tracking: b/233647273

**Comment by chsigg** [comment]
Fixed with https://github.com/tensorflow/tensorflow/commit/6d6d7fece1752523699993d7a30ee7790e84adba. I will leave it to Jake to close this issue.

**Comment by jakevdp** [comment]
Thanks!

**Comment by chsigg** [comment]
Unfortunately, the change got rolled back. I will resubmit as soon as the failure that it triggered is resolved. I'm hoping it's only a test that is too restrictive.

**Comment by jakevdp** [comment]
Looks like this is fixed in the jaxlib 0.3.14 release
